### PR TITLE
Add health checks for Event Hub component

### PR DESCRIPTION
**Pull request by oising**
Add healthchecks for eventhub component

closes https://github.com/dotnet/aspire/issues/3980

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4050)

### CODE REVIEW

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs [hunk 1]
```diff
@@ -101,4 +103,47 @@ protected static string GetNamespaceFromSettings(AzureMessagingEventHubsSettings
                 $"the '{configurationSectionName}' configuration section.");
         }
     }
+    protected sealed class AzureEventHubHealthCheck : IHealthCheck, IAsyncDisposable
```
Comment (danmoseley) [inline src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs:106]: nit, newline above

actually also src/Components/** don't seem to be completely consistent but in general seem to prefer only one non-nested class per file. Eg `SeqHealthCheck` is its own file, `NatsHealthCheck` etc.

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs [hunk 2]
```diff
@@ -101,4 +103,47 @@ protected static string GetNamespaceFromSettings(AzureMessagingEventHubsSettings
...
+        public AzureEventHubHealthCheck(TSettings settings)
+        {
+            var probeOptions = new EventHubProducerClientOptions
+            {
+                Identifier = $"AspireEventHubHealthCheck-{settings.EventHubName}",
+                RetryOptions = new EventHubsRetryOptions()
+                {
+                    Mode = EventHubsRetryMode.Exponential,
+                    MaximumRetries = 3,
+                    Delay = TimeSpan.FromMilliseconds(800),
```
Comment (danmoseley) [inline src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs:119]: where is 800 from?

Comment (oising) [reply src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs:119]: I'll hoist this into Aspire:Settings :)

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs (hunk 1 above)
Comment (oising) [reply src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs:106]: I get you, but it's only used in the derivations of EventHubsComponent itself, so I find private-protected seems to be the right visibility in this case.

**Comment by oising** [comment]
image

If you look at the eventhubs playground API project (it's the only webapp based one capable of exposing /health) you can trace the logic. 

I also added a DisableHealthChecks setting to align with other components.

**Comment by joperezr** [comment]
@oising thanks for the PR and for your contribution. I noticed that your PR is targeting 8.0 branch, which we are no longer taking changes for. Is there a specific reason why you picked `release/8.0` as the target branch and would you be okay with us retargeting to `main` branch instead?

**Comment by oising** [comment]
> @oising thanks for the PR and for your contribution. I noticed that your PR is targeting 8.0 branch, which we are no longer taking changes for. Is there a specific reason why you picked `release/8.0` as the target branch and would you be okay with us retargeting to `main` branch instead?

Retarget as you please!

**Comment by joperezr** [comment]
Actually, I was about to retarget, but because this has changes that are only on release/8.0 branch that we specifically don't want in main, things wouldn't go so well. Due to that, I think the best way would be to reset and cherry-pick the changes directly into main's tip, and then update this PR. Want me to help with that?

File: src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs [hunk 3]
```diff
@@ -46,6 +46,11 @@ public abstract class AzureMessagingEventHubsSettings : IConnectionStringSetting
     /// </summary>
     public TokenCredential? Credential { get; set; }
 
+    /// <summary>
+    /// Gets or sets a boolean value that indicates whether the health checks are disabled or not.
+    /// </summary>
```
Comment (sebastienros) [inline src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:51]:
```suggestion
    /// <summary>
    /// Gets or sets a boolean value that indicates whether the Azure Event Hubs health check is disabled or not.
    /// </summary>
    /// <value>
    /// The default value is <see langword="false"/>.
    /// </value>
```

Comment (sebastienros) [reply src/Components/Aspire.Azure.Messaging.EventHubs/AzureMessagingEventHubsSettings.cs:51]: I know it's a lame request but it makes everything the same. Will also require updating the schema files.

File: src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs (hunk 2 above)
Comment (sebastienros) [reply src/Components/Aspire.Azure.Messaging.EventHubs/EventHubsComponent.cs:122]: @eerhardt I don't think there is a precedent for health check retry options. I found `MongoDBSettings.HealthCheckTimeout` only.

Would adding a `RetryOptions HealthCheckRetryOptions` property in `AzureMessagingEventHubsSettings` work here?

**Comment by oising** [comment]
Created new PR targeting main (cherry pick) at https://github.com/dotnet/aspire/pull/4139

/cc @joperezr @danmoseley @sebastienros

**Pull request closed without merging**